script.sh

```diff
@@ -208,14 +210,45 @@ RUN echo "deb-src ${TRAVIS_DEBIAN_MIRROR} experimental main" >> /etc/apt/sources
 EOF
 fi
 
+case "${TRAVIS_EXTRA_REPOSITORY}" in
+	https:*)
+		EXTRA_PACKAGES="apt-transport-https"
+		;;
+esac
+
+if [ -n "${TRAVIS_EXTRA_REPOSITORY_GPG_URL}" ]
+then
+	EXTRA_PACKAGES="${EXTRA_PACKAGES} wget gnupg"
+fi
+
 cat >>Dockerfile <<EOF
 RUN apt-get update && apt-get dist-upgrade --yes
-RUN apt-get install --yes --no-install-recommends build-essential equivs devscripts git-buildpackage ca-certificates pristine-tar
+RUN apt-get install --yes --no-install-recommends build-essential equivs devscripts git-buildpackage ca-certificates pristine-tar ${EXTRA_PACKAGES}
 
 WORKDIR $(pwd)
 COPY . .
 EOF
 
+if [ -n "${TRAVIS_EXTRA_REPOSITORY_GPG_URL}" ]
+then
+	cat >>Dockerfile <<EOF
+RUN wget -O- "${TRAVIS_EXTRA_REPOSITORY_GPG_URL}" | apt-key add -
+EOF
+fi
+
+# We're adding the extra repository only after the essential tools have been
+# installed, so that we have apt-transport-https if the repository needs it.
+# The subsequent apt-get update also avoids useless re-downloading of Debian
+# package lists we already have.
+if [ -n "${TRAVIS_EXTRA_REPOSITORY}" ]
+then
+	cat >>Dockerfile <<EOF
+RUN echo "deb ${TRAVIS_EXTRA_REPOSITORY}" >> /etc/apt/sources.list.d/travis_extra.list
+RUN echo "deb-src ${TRAVIS_EXTRA_REPOSITORY}" >> /etc/apt/sources.list.d/travis_extra.list
+RUN apt-get update -o Dir::Etc::sourcelist=/etc/apt/sources.list.d/travis_extra.list -o Dir::Etc::sourceparts=- --no-list-cleanup
```

Comment:
Those parameters make `apt-get update` act only on the just added extra repository, so that we avoid useless re-downloading of the Debian package lists we already have. I've extended the script comment to include what each of the arguments is doing.